packages/rest/src/rest.server.ts

```diff
@@ -180,6 +180,13 @@ export class RestServer extends Context implements Server, HttpServerLike {
       config.host = undefined;
     }
 
+    this.basePath(config.basePath);
+
+    const OPENAPI_SPEC_MAPPING: {[key: string]: OpenApiSpecForm} = {
+      [`${this._basePath}/openapi.json`]: {version: '3.0.0', format: 'json'},
+      [`${this._basePath}/openapi.yaml`]: {version: '3.0.0', format: 'yaml'},
```

Comment:
I am also not sure we want these endpoints to honor `basePath` option.

@raymondfeng @strongloop/loopback-maintainers what's your opinion? When the application is configured with a custom `basePath`, e.g. `/api`, should we apply the base path to openapi endpoints too?

- Now we have `/openapi.json` and e.g. `/api/products`
- Proposed change: `/api/openapi.json` and e.g. `/api/products`

I am afraid the proposed implementation is a breaking change requiring a semver-major release. I think we should preserve backwards compatibility and introduce a new feature flag allowing users to decide whether to apply basePath configuration to openapi endpoints or not. By default, we should fall back to old behavior, but new projects scaffolded by `lb4 cli` should enable the new behavior.

Reply:
@bajtos Having /explorer always running running at the root creates a problem when there's a reverse proxy running in front of the API. Here's our situation and I'm sure this is pretty common case.
 
We have both UI and API applications running on the same host behind `nginx`. 
 
UI runs at: `http://www.mysite.com`
Loopback API runs at: `http://www.mysite.com/_api`

Whenever a requests come for `/_api/*` `nginx` forwards it to Loopback API. All other requests to to the UI application.  
 
As a result requests http://www.mysite.com/explorer fail because they don't even reach the Loopback app. Hence my fix, which allows explorer to run at http://www.mysite.com/_api/explorer .

Reply:
I don't think `basePath` was introduced for this purpose. There are 3 paths that build a route to a controller method:

1. server-level `basePath`: it's used as `_expressApp.use(basePath, <lb4-rest-router>);`, such as `/api`
2. controller-class-level `basePath` (set by @api): it's prepended to the path from step 3, such as `/products`
3. controller-method-level `path` (set by @get, @post etc): local path of the operation, such as `/{id}`

Please note 1 is for express middleware mounting and it is transparent to our REST API routes. Express will strip `/api` from `request.url` property, which becomes `/products/{id}`.
2 + 3 is for internal routing by `@loopback/rest`.

Putting reverse proxy aside, the http url for a method is: `/<serverBasePath>/<controllerBasePath>/<operationPath>`, such as `/api/products/{id}`and api explorer is exposed at `/explorer` by default.

Reply:
@raymondfeng Then my question is: is there a way to expose the API explorer at some different route (like `/myprefix/explorer`)?

I did create a workaround in my project but it's quite hacky and involves messing with the express route stack directly. I'd like to avoid doing it.

Reply:
@dkrantsberg I'd like to better understand your scenario.

> We have both UI and API applications running on the same host behind nginx.
> 
> - UI runs at: http://www.mysite.com
> - Loopback API runs at: http://www.mysite.com/_api

Do you have two servers, one for UI and another for LoopBack API? If that's the case, then I would expect you should be able to configure your nginx instance as follows:
 - API requests (e.g. `http://www.mysite.com/_api/products/{id}`) are forwarded to LB instance and `_api` is stripped from the path (e.g. `/products/{id}`).
- As a result, explorer will become available at `http://www.mysite.com/_api/explorer`. Alternatively, I believe it should be possible to configure nginx and define a new virtual route (e.g. `http://www.mysite.com/api-explorer`) that will be forwarded to `/explorer` in LB.

I think the key is to keep empty `basePath` setting in LoopBack.

In case you are running both UI and LB API in the same server, how do you compose these two sets of routes?

Reply:
> is there a way to expose the API explorer at some different route (like `/myprefix/explorer`)?

The self-hosted version of API Explorer that's provided by `@loopback/rest-explorer` is already honoring `basePath` setting and allows customization of the Explorer URL, see 
https://github.com/strongloop/loopback-next/tree/master/packages/rest-explorer#basic-use

> By default, API Explorer is mounted at /explorer. This path can be customized via RestExplorer configuration as follows:
> ```ts
> this.bind(RestExplorerBindings.CONFIG).to({
>   path: '/openapi/ui',
> });
> ```

The redirection to externally-hosted API Explorer, which is built into `@loopback/rest`, does not support customization of the URL and it also ignores `basePath` setting. See here:
https://github.com/strongloop/loopback-next/blob/6e0eeb66b6a8c79d736fcc2dc7d9c76bfd57abd3/packages/rest/src/rest.server.ts#L288-L291

Please note that these redirects are deprecated, we want our users to use the self-hosted variant provided by `@loopback/rest`.